packages/core/src/utils.ts

```diff
@@ -705,24 +713,112 @@ export const getParentContractASTNodes = (
   return parentContractNodes
 }
 
+/**
+ * Borrowed from the Hardhat ethers plugin. Unfortunately, it does not export this function.
+ * Since we had to copy this function into our code base anyway, we've also modified it to
+ * move all of the validation into the chugsplash config parsing and validation logic.
+ *
+ * So in this function, we assume all the libraries are valid.
+ * Source: https://github.com/NomicFoundation/hardhat/blob/5426f582c28f90576300378c94a86e61ca13bc7b/packages/hardhat-ethers/src/internal/helpers.ts#L163
+ *
+ * @param artifact The artifact of the contract we are fetching the bytecode of.
+ * @param libraries The libraries to link into the bytecode.
+ * @returns The bytecode of the contract with the libraries linked in.
+ */
+const getBytecodeWithExternalLibraries = async (
+  artifact: ContractArtifact,
+  libraries: Libraries
+) => {
+  const neededLibraries: Array<{
+    sourceName: string
+    libName: string
+  }> = []
+  for (const [sourceName, sourceLibraries] of Object.entries(
+    artifact.linkReferences
+  )) {
+    for (const libName of Object.keys(sourceLibraries)) {
+      neededLibraries.push({ sourceName, libName })
+    }
+  }
+  const linksToApply: Map<string, Link> = new Map()
+  for (const [linkedLibraryName, linkedLibraryAddress] of Object.entries(
+    libraries
+  )) {
+    const matchingNeededLibraries = neededLibraries.filter((lib) => {
+      return (
+        lib.libName === linkedLibraryName ||
+        `${lib.sourceName}:${lib.libName}` === linkedLibraryName
+      )
+    })
+
+    const [neededLibrary] = matchingNeededLibraries
+
+    const neededLibraryFQN = `${neededLibrary.sourceName}:${neededLibrary.libName}`
+
+    linksToApply.set(neededLibraryFQN, {
+      sourceName: neededLibrary.sourceName,
+      libraryName: neededLibrary.libName,
+      address: linkedLibraryAddress,
+    })
+  }
+
+  return linkBytecode(artifact, [...linksToApply.values()])
+}
+
+/**
+ * Links the given bytecode with the given libraries.
+ * Borrowed from hardhat ethers plugin and updated to fit our conventions.
+ * Source: https://github.com/NomicFoundation/hardhat/blob/5426f582c28f90576300378c94a86e61ca13bc7b/packages/hardhat-ethers/src/internal/helpers.ts#L424
+ *
+ * @param artifact The artifact of the contract that will be linked.
+ * @param libraries The libraries to link.
+ * @returns The linked bytecode.
+ */
+const linkBytecode = (
```

Comment:
I think this should be exported. Might wanna check the rest of the functions in this PR to see if we missed any that should be exported

Reply:
Hardhat probably thought the same thing when they didn't export `[collectLibrariesAndLink](https://github.com/NomicFoundation/hardhat/blob/5426f582c28f90576300378c94a86e61ca13bc7b/packages/hardhat-ethers/src/internal/helpers.ts#L163)`, which prevented us from using it even if we wanted to